### ISSUE TITLE
Investigate possible performance/speed improvements

### ISSUE THREAD

**Issue by sideshowbarker** [tabatkins/highlighter#6]
The normal HTML spec build time is on the order of 15 seconds — that is, without the highlighter.

As discussed on \#whatwg IRC around https://freenode.logbot.info/whatwg/20180607#c1579082, running the HTML spec build with the highlighter incorporated causes the build time to increases to more than 4 minutes.

While it’s true that that highlighter is called multiple times for each `<pre>`, that ~4 minutes of time indicates it takes about 80 milliseconds to process each `<pre>` once.

We have 1121 `<pre>` elements in the HTML spec, so if we were to call the highlighter on each one only once, that would take about 90 seconds total.

So the highlighter integration is adding ~90 seconds to the build time (for each spec-version build).

To isolate where that time is being spent, I did some simple investigation with wattsi by replacing the call to the highlighter in the wattsi code with just a call to `/bin/cat`.

The resulting build time was 29 seconds. That is, it increased by just 13 seconds (from 16 to 29).

So that seems to indicate that only 13 seconds are spent executing the code I added to wattsi for extracting all the examples and IDL blocks from the spec and piping each to an external process and getting that resulting output back and writing it to the generated spec HTML.

And that suggests the bulk of the ~90 additional seconds needed to run the highlighter is spent in the highlighter code itself — and not in the I/O on the wattsi side or the internal wattsi code.

Therefore it might be useful to investigate whether any performance/speed improvements can be made in the highlighter code.

**Comment by tabatkins** [comment]
I'm surprised that IO was so little, and the highlighting itself was comparatively so much.  That said, I've done absolutely zero perf-hacking on this so far, and am happy to try this out. I'll adapt the Bikeshed profiling machinery to see what I can tighten up.

**Comment by sideshowbarker** [comment]
Sounds great — I’m happy to help with testing or in any other way needed

**Comment by tabatkins** [comment]
So I'm running all ~1100 samples from <https://gist.githubusercontent.com/sideshowbarker/8284404/raw/8efdf8fe9713839b839955c9a099ff90e2e559a1/examples.json>, which should be the entirety of HTML, and printing the results to the console, and the entire things takes about 2.3s:

```
real	0m2.289s
user	0m2.216s
sys	0m0.056s
```

This is in line with what I thought should happen, which suggests that the actual slowdown is somewhere else. I suspect that it's in freshly invoking the Python interpreter 1100 times.

Instead of running `/bin/cat`, could you just make an empty .py file somewhere with the following contents:

```python
#!/usr/bin/env python2.7
# -*- coding: utf-8 -*-
pass
```

And invoke *that* instead of highlighter?  I'll bet that should still take the majority of the time you're seeing.

If that's the case, then the only thing we can do to improve things is batch the calls together into a single script invocation. Is that possible to do in the Wattsi pipeline?

**Comment by tabatkins** [comment]
And confirmed, I wrote another test that shelled out to highlight for each line, and it's taking *much* longer. Having trouble getting `time` to spit out something useful, but it's definitely at least a minute.

**Comment by sideshowbarker** [comment]
> If that's the case, then the only thing we can do to improve things is batch the calls together into a single script invocation. Is that possible to do in the Wattsi pipeline?

No — while anything’s possible, that would not at all be practical to do it in Wattsi in its current form

**Comment by tabatkins** [comment]
Okay, I just uploaded some new code that might work.  Instead of invoking `highlighter/__init__.py`, could you instead invoke `highlighter/server.py &` (the `&` will put it into the background), then `curl -g localhost:8080/lang-goes-here?json-goes-here`? It *should* spit out the desired html string.

For example, `curl -g http://localhost:8080/webidl?[%22pre%22,%7B%22class%22:%22idl%22%7D,%22interface%20Foo%20{%20readonly%20attribute%20long%20bar;%20};%22]` responds with `<pre class='idl'><c- b>interface</c-> <c- g>Foo</c-> { <c- b>readonly</c-> <c- b>attribute</c-> <c- b>long</c-> <c- g>bar</c->; };</pre>`.

Note that you'll have to url-encode the json or else the server gets angry.

Hopefully this should be much faster, since it only starts up Python once. You'll be paying some HTTP overhead, but it's all local, so it shouldn't be too expensive.

**Comment by sideshowbarker** [comment]
Thanks extremely much. I hooked the highligh server into the wattsi pipeline and it worked perfectly.

It’s just an initial interim integration using `curl` as described above — but as expected, much much faster than re-calling the script each time.

As far as how fast it is, it takes the build time from the 5 minutes and 45 seconds it was requiring, and takes it down to about 1 minute and 45 seconds. That is, so far it’s reduced the time by 4 minutes.

While 105 seconds is still a big jump up from the 16 seconds the build was taking without this, it is worth remembering that the build isn’t doing the highlighting just once — instead it’s doing it 4 times.

So that means it’s taking about 23 seconds for each pass of the 1121 `pre` elements — or about 20 milliseconds on average for each `pre`.

But in this initial implementation, that has the cost of using shell process I/O by invoking `curl` to make the request to the server API and get the response back.

It seems like it will be faster to instead replace that shell-process invocation with compiled code that does the same thing using the FPC native internal `fphttpclient` http://wiki.freepascal.org/fphttpclient to handle the request and response.

So next I’ll try writing that up and see what we get.

**Comment by tabatkins** [comment]
Out of curiousity, is there any way for you to cache the highlight results across runs, so that you only actually call into highlighter for the first run, and all subsequent runs can just immediately get their highlight results back? That should be ~free for the following runs, dropping 75% of the runtime.

**Comment by sideshowbarker** [comment]
> Out of curiousity, is there any way for you to cache the highlight results across runs

There’s at least no very-easy way to do that from that wattsi as far as I can see. I had hoped maybe the Free Pascal HTTP client had caching support built-in, but apparently it doesn’t.

Of course I could write some custom code to do caching but given that hooking the highlight server into the pipeline has reduced the additional build time for syntax highlighting to only 11 seconds, I think it’s not worth the time to try to implement caching now. I can investigate it more later (after the initial non-caching implementation gets landed).

**Comment by sideshowbarker** [comment]
> Out of curiousity, is there any way for you to cache the highlight results across runs

Yeah, I just now went ahead and added a simple caching mechanism to the Wattsi code. Adding that shaved an additional two seconds off the time needed for the syntax-highlighting build. So the total build time with syntax highlighting enabled is now 25 seconds in my environment — compared to 16 seconds without syntax highlighting enabled. So the performance cost is just 9 seconds total.

It’s imaginable there’s a faster way to do the caching than the way I did it. But we’re at the point where any additional speedup it might bring be wouldn’t be noticeable enough in practice to matter much.
